### PR TITLE
[PNUDEV-30] Add pen name field to writer account

### LINKED ISSUE #30: Add pen name field to writer account
1. Add pen name field to writer account.
2. Add ability for writer to decide how to sign his posts: using pen name or using full name (checkbox on edit account page).

### PR DESCRIPTION
**Pull request by volodymyrvynnyk**
### Add pseudonym to PublicAccount and fixed assertion exception messages in tests

**When Admin filter posts by author, they now see pseudonym, if author uses pseudonym**
image

**User-Writer can choose options that hide their name and show pseudonym**
image

### CODE REVIEW

File: src/main/java/com/pnu/dev/shpaltaif/domain/PublicAccount.java [hunk 1]
```diff
@@ -36,6 +39,10 @@
     @NotNull
     private String surname;
 
+    private String pseudonym;
+
+    private boolean pseudonymUsed;
```
Comment (Igor-Polatajko) [inline src/main/java/com/pnu/dev/shpaltaif/domain/PublicAccount.java:44]: usePseudonym will be a better naming

File: src/main/java/com/pnu/dev/shpaltaif/dto/PublicAccountDto.java [hunk 2]
```diff
@@ -14,11 +16,19 @@
...
 
     @NotNull(message = "Ім'я повинно бути вказаним")
+    @NotBlank
+    @Length(min = 3)
     private String name;
 
     @NotNull(message = "Прізвище повинно бути вказаним")
+    @NotBlank
+    @Length(min = 3)
     private String surname;
```
Comment (Igor-Polatajko) [inline src/main/java/com/pnu/dev/shpaltaif/dto/PublicAccountDto.java:23]: Remove NotNull and move message to NotBlank

File: src/main/java/com/pnu/dev/shpaltaif/service/PublicAccountServiceImpl.java [hunk 3]
```diff
@@ -66,11 +71,26 @@ public PublicAccount create(PublicAccountDto publicAccountDto, User user) {
...
 
+        if (nonNull(publicAccountDto.getPseudonym())) {
+            publicAccountDto.setPseudonym(publicAccountDto.getPseudonym().trim());
+            if (publicAccountRepository.existsByPseudonym(publicAccountDto.getPseudonym())) {
+                throw new ServiceException("Псевдонім зайнятий");
+            }
+        }
+
+        if (publicAccountDto.isPseudonymUsed()) {
+            if (isNull(publicAccountDto.getPseudonym()) || StringUtils.isBlank(publicAccountDto.getPseudonym())) {
```
Comment (Igor-Polatajko) [inline src/main/java/com/pnu/dev/shpaltaif/service/PublicAccountServiceImpl.java]: I believe checking for isBlank() will be enough
Also, avoid nested "if"s

File: src/test/java/com/pnu/dev/shpaltaif/CategoryServiceIntegrationTest.java [hunk 4]
```diff
@@ -129,10 +129,9 @@ public void findByIdNotFound() {
 
         assertEquals(0, categoryService.findAll().size());
 
-        assertThrows(ServiceException.class,
-                () -> categoryService.findById(Long.MAX_VALUE),
-                "Категорію не знайдено"
-        );
+        ServiceException thrown = assertThrows(ServiceException.class,
```
Comment (Igor-Polatajko) [inline src/test/java/com/pnu/dev/shpaltaif/CategoryServiceIntegrationTest.java:132]: Do you think the library method doesn't work the way is supposed to?

File: src/test/java/com/pnu/dev/shpaltaif/PublicAccountServiceIntegrationTest.java [hunk 5]
```diff
@@ -128,4 +133,65 @@ public void createAndThenReadAndThenUpdateAndThenDeleteSuccessFlow() {
...
+                .pseudonymUsed(true)
+                .pseudonym("Pseudonym")
+                .build();
+
+        thrown = assertThrows(ServiceException.class,
+                () -> publicAccountService.update(publicAccountDtoWithExistsPseudonym, publicAccount2.getId()));
+        assertEquals(thrown.getMessage(), "Псевдонім зайнятий");
+    }
+
+    private User createUserWriter(String username) {
```
Comment (Igor-Polatajko) [inline src/test/java/com/pnu/dev/shpaltaif/PublicAccountServiceIntegrationTest.java]: I'd suggest testing the following flow instead:
1. Create and save user and account (I'd use userRepository for dealing with user).
2. Check that publicAccount.getSignature() is what you expect.
3. Update account and check publicAccount.getSignature() again.

And another test for incorrect signature and duplicate username.

I'd recommend using userRepository instead of userService, because here you use side effect of user creation, that creates empty account, but actually it's the piece of logic that have to be tested by tests for user creation flow.

**Review by Igor-Polatajko — CHANGES_REQUESTED** [review]

File: src/main/java/com/pnu/dev/shpaltaif/domain/PublicAccount.java (hunk 1 above)
Comment (volodymyrvynnyk) [reply src/main/java/com/pnu/dev/shpaltaif/domain/PublicAccount.java:44]: I thought that way
But getter method will be named as isUsePseudonym, it's incorrect at grammar point
So, that's why I decided to name it pseudonymUsed -> isPseudonymUsed

File: src/test/java/com/pnu/dev/shpaltaif/CategoryServiceIntegrationTest.java (hunk 4 above)
Comment (volodymyrvynnyk) [reply src/test/java/com/pnu/dev/shpaltaif/CategoryServiceIntegrationTest.java:132]: Yes, I do think
I've checked whether it works correct in few Test classes, so, each test class will passed correctly, even if messages are "any text".
And I've also googled the approach how to test exception message, and this is correct way

File: src/test/java/com/pnu/dev/shpaltaif/PublicAccountServiceIntegrationTest.java [hunk 6]
```diff
@@ -134,64 +129,108 @@ public void createAndThenReadAndThenUpdateAndThenDeleteSuccessFlow() {
...
+                .pseudonym(pseudonym)
                 .build();
+        publicAccountService.update(publicAccountDto, publicAccountFromDb.getId());
 
-        ServiceException thrown = assertThrows(ServiceException.class,
-                () -> publicAccountService.update(publicAccountDtoWithoutPseudonym, publicAccount1.getId()));
-        assertEquals(thrown.getMessage(), "Щоб використовувати псевдонім, введіть його коректно");
+        //Test signature when pseudonym has been set and pseudonym is active
+        publicAccountFromDb = publicAccountService.findById(publicAccount.getId());
+        expectedSignature = pseudonym.trim();
```
Comment (Igor-Polatajko) [inline src/test/java/com/pnu/dev/shpaltaif/PublicAccountServiceIntegrationTest.java]: Do not reassign variables, it's really become hard to read.
Just create a new variable each time.
Or use value directly in assertEquals(), because it's clear enough, that first arg have to be an expected result.

File: src/test/java/com/pnu/dev/shpaltaif/PublicAccountServiceIntegrationTest.java [hunk 7]
```diff
@@ -134,64 +129,108 @@ public void createAndThenReadAndThenUpdateAndThenDeleteSuccessFlow() {
...
+                .surname(surname)
                 .build();
 
-        userService.create(createUserDto);
-        List<User> allUsers = userRepository.findAll();
-        assertEquals(usersNumberBeforeCreate + 1, allUsers.size());
+        return publicAccountService.create(publicAccountDto, user);
+    }
+
+    private User createUserWriter() {
```
Comment (Igor-Polatajko) [inline src/test/java/com/pnu/dev/shpaltaif/PublicAccountServiceIntegrationTest.java]: no need in this method, just move this code to the method above.
I'd even suggest moving this code to the tests methods itself, along with account creation (actually, it's not a duplicate, it's test preparation, and the simplier it will look like, the better it will be).

File: src/test/java/com/pnu/dev/shpaltaif/PublicAccountServiceIntegrationTest.java [hunk 8]
```diff
@@ -134,64 +129,108 @@ public void createAndThenReadAndThenUpdateAndThenDeleteSuccessFlow() {
...
+        int usersNumberBeforeCreate = userRepository.findAll().size();
+        long userId = usersNumberBeforeCreate + 1;
+        User user = User.builder()
+                .id(userId)
```
Comment (Igor-Polatajko) [inline src/test/java/com/pnu/dev/shpaltaif/PublicAccountServiceIntegrationTest.java]: do not use this userId, just skip this field.
I'd suggest to follow the comment above, but in opposite case, just set a random username.
No need in this calculations

**Review by Igor-Polatajko — CHANGES_REQUESTED** [review]

File: src/main/resources/db/migration/V5__AddedPseudonymToPublicAccount.sql [hunk 9]
```diff
@@ -0,0 +1,5 @@
+ALTER TABLE public_account
```
Comment (Igor-Polatajko) [inline src/main/resources/db/migration/V5__AddedPseudonymToPublicAccount.sql]: Let's make version name to start with V<current_version>.0.<branch_name>
e. g. V5.0.37__

It will reduce feature merging conflicts.

**Review by Igor-Polatajko — APPROVED** [review]
Good!
One comment regarding migrations.
